File: problems/Swift/6.Z字形变换.swift
/**
 * 6.Z字形变换 [Medium]
 * Time Complexity: O(n), Space Complexity: O(n)
 * https://leetcode-cn.com/problems/zigzag-conversion/
 */

class Solution {
  func convert(_ s: String, _ numRows: Int) -> String {
    guard numRows > 1 else {
      return s
    }

    let chars = Array(s)
    var result: [Character] = []
    var cycleLen = 2 * numRows - 2

    for i in 0 ..< numRows {
      var j = 0
      while j + i < chars.count {
        result.append(chars[j + i])
        if i != 0, i != numRows - 1, j + cycleLen - i < chars.count {
          result.append(chars[j + cycleLen - i])
        }

        j += cycleLen
      }
    }

    return String(result)
  }
}

File: problems/rust-solutions/src/6.Z字形变换.rs
// 6.Z字形变换 [Medium]
// Time Complexity: O(n), Space Complexity: O(n)
// https://leetcode-cn.com/problems/zigzag-conversion/

impl Solution {
  pub fn convert(s: String, num_rows: i32) -> String {
    if num_rows == 1 {
      return s;
    }

    let mut result = String::new();
    let cycleLen = 2 * num_rows - 2;
    let chars = s.as_bytes();

    for i in 0..num_rows {
      let mut j = 0;
      while j + i < s.len() as i32 {
        result.push(chars[(j + i) as usize] as char);
        if i != 0 && i != num_rows - 1 && j + cycleLen - i < s.len() as i32 {
          result.push(chars[(j + cycleLen - i) as usize] as char)
        }
        j += cycleLen;
      }
    }

    return result;
  }
}
